File: app/src/main/java/com/fractalwrench/crazycats/injection/DefaultSchedulers.java
package com.fractalwrench.crazycats.injection;

import io.reactivex.Observable;
import io.reactivex.ObservableTransformer;
import io.reactivex.Scheduler;

/**
 * Defines io/ui schedulers (which are swapped out to facilitate testing)
 */
public class DefaultSchedulers {

    private final Scheduler io;
    private final Scheduler ui;

    public DefaultSchedulers(Scheduler io, Scheduler ui) {
        this.io = io;
        this.ui = ui;
    }

    public Scheduler io() {
        return io;
    }

    public Scheduler ui() {
        return ui;
    }

    /**
     * Should be called using {@link Observable#compose(ObservableTransformer)} to set the default
     * schedulers for an observable.
     *
     * @param observable the observable
     * @param <T>        type
     * @return the observable with newly set schedulers
     */
    public <T> Observable<T> apply(Observable<T> observable) {
        return observable.subscribeOn(io)
                         .observeOn(ui);
    }

}

File: app/src/main/java/com/fractalwrench/crazycats/injection/DefaultSchedulers.kt
package com.fractalwrench.crazycats.injection

import io.reactivex.Observable
import io.reactivex.ObservableTransformer
import io.reactivex.Scheduler

/**
 * Defines io/ui schedulers (which are swapped out to facilitate testing)
 */
class DefaultSchedulers(private val io: Scheduler, private val ui: Scheduler) {

    fun io(): Scheduler {
        return io
    }

    fun ui(): Scheduler {
        return ui
    }

    /**
     * Should be called using [Observable.compose] to set the default
     * schedulers for an observable.

     * @param observable the observable
     * *
     * @param <T>        type
     * *
     * @return the observable with newly set schedulers
    </T> */
    fun <T> apply(observable: Observable<T>): Observable<T> {
        return observable.subscribeOn(io)
                .observeOn(ui)
    }

}
